### PR TITLE
Custom Elytras

### PR DESCRIPTION
**Pull request by Technici4n**
Reopening #1374 already. This is not final, I have to investigate if this addition will also work for non-player living entities, and document that behavior.

### CODE REVIEW

**Comment by Technici4n** [comment]
Updated with a remark, I checked this on a dedicated server and there was no issue, and I checked that it redirected every necessary call in the past. Once you agree, it can be marked as `merge-me-please` @liach @Juuxel.

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
also quick question, how has the api's content mechanism changed?

**Comment by OroArmor** [comment]
@Technici4n , after recent events, i would prefer if this pull request did not have me attached to it. thank you

**Comment by Technici4n** [comment]
> @Technici4n , after recent events, i would prefer if this pull request did not have me attached to it. thank you

@OroArmor Want me to squash and remove you from the authors? Would that be OK?

**Comment by OroArmor** [comment]
Yeah. That works

**Comment by Technici4n** [comment]
First time using interactive rebase, I believe this is correct (the diff is empty).

**Comment by Technici4n** [comment]
Needs an update to broadcast an `ALLOW_ELYTRA_FLIGHT` event, and use #1369 to allow easy implementation on items. There are lots of mods that could want to expose elytra flight behavior in other ways, such as Trinkets or a mod adding custom potions, etc...

**Comment by Technici4n** [comment]
Ok, this is the result of another investigation for all uses of `Items.ELYTRA`:
* `ModelPredicateProviderRegistry#register`, specific to the elytra item.
* `ClientPlayerEntity#tickMovement`, responsible for starting the elytra flight. Need a hook there.
* `CapeFeatureRenderer` to stop capes from rendering when an elytra is worn. No need for an elytra hook there imo, we can have a generic `ALLOW_CAPE_RENDER` later and let modders use mixin for now.
* `ElytraFeatureRenderer` to render the elytra. This is specific to the elytra item, other modders can either hook there or register their own renderer.
* `EndTabAdvancementGenerator` to generate the advancement. No hook needed.
* `LivingEntity#onEquipStack` for the equip sound - I think we already have a hook for this. In all cases, no hook needed for elytra flight.
* `LivingEntity#initAi` for the actual elytra motion. This handles both the flight continuation check and the damage of the item. Need a hook there.
* `MobEntity#getPreferredEquipmentSlot` - no need for a hook.
* `PlayerEntity#checkFallFlying` - used by `ClientPlayerEntity#tickMovement` to start elytra flight, we can group both into one event dispatch.
* `EndCityGenerator.Piece` - no need for a hook.

TL;DR:
* We need an `ALLOW_ELYTRA_FLIGHT_START` **client-only** event to start the flight. **`ClientPlayerEntity` only.**
* We need a `CONTINUE_ELYTRA_FLIGHT` event on both the client and the server to continue the flight, and handle the damage of the item (or mana or whatever) **server-side only**. We will have to more or less overwrite the vanilla `initAi` function.  **Any `LivingEntity`.**

**Comment by Technici4n** [comment]
And I just checked that LivingEntities can be fallflying, but won't enable it for themselves automatically. It can be activated with commands though (or by another mod), will be worth adding to the javadoc.

https://www.reddit.com/r/minecraftsuggestions/comments/6crat3/mobs_that_fall_while_wearing_an_elytra/

**Comment by Technici4n** [comment]
Going to open a new one actually.
